### PR TITLE
Initial commit; REST APIs and basic kubectl ability

### PR DESCRIPTION
**Pull request by ct7t**

### CODE REVIEW

**Review by josh-7t — APPROVED** [review]
Clean!

File: src/k8s/core.clj [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+             {:body ((comp keybab-keys json/read-value) (:body response))}))))
+
+(defn- api
+  [k op]
+  (when-let [url (k api-map)]
+    (if (= http/post op)
+      url
+      (str url "{{name}}"))))
+
+(defn- camel-keys
```
Comment (joel7t) [inline src/k8s/core.clj:53]: Could this be replaced by `(comp inflections/stringify-keys inflections/camel-case-keys)`?  I'm not sure if the order is correct in the comp, and maybe `inflections/str-name` will be better than `inflections/stringify-keys`?

If there is a specific behaviour that you need that's different from what inflections provides, maybe add a docstring to capture those differences.

https://cljdoc.org/d/inflections/inflections/0.13.2/api/inflections.core#camel-case-keys

**Review by joel7t — APPROVED** [review]

File: src/k8s/kubectl.clj [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+(defn- k8s-rest
+  [f uri {:keys [kind metadata] :as request}]
+  (let [response (f uri request)
+        status (-> response :status)]
+    (cond
+      (nil? status) response
+      (and (<= status 299)
+           (>= status 200)) (-> response :body)
+      :else (throw (Exception. (-> response :body :message))))))
```
[the remark this answers is not in the document]
Comment (ct7t) [reply src/k8s/kubectl.clj:36]: The intent here is that if you're interacting with `kubectl` you're abstracted from the fact that there's an underlying HTTP connection at play (just like it is on the command line).  So, it's a deliberate choice to withhold the status and just return you the message or body

(unless no status can be found - which means this isn't an HTTP response so the code doesn't know what to do and escalates)

File: src/k8s/kubectl.clj [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  (k8s-rest k8s/get uri request))
+
+(defn apply!
+  [uri request]
+  (try
+    (create! uri request)
+    (catch Exception e (replace! request))))
+
+(defn- k8s-rest
+  [f uri {:keys [kind metadata] :as request}]
```
[the remark this answers is not in the document]
Comment (ct7t) [reply src/k8s/kubectl.clj:29]: Good catch.  Thanks.  Will fix.

File: src/k8s/kubectl.clj [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+(defn get
+  [uri request]
+  (k8s-rest k8s/get uri request))
+
+(defn apply!
+  [uri request]
+  (try
+    (create! uri request)
+    (catch Exception e (replace! request))))
```
[the remark this answers is not in the document]
Comment (ct7t) [reply src/k8s/kubectl.clj:26]: These are good questions.  The intent of `apply!` is basically an upsert.  More grace will be needed here, and there are a few object types where the rules are slightly different.  This is a big "covers 80% of cases, and 95% of cases I care about" hammer.  It's definitely my goal to evolve this over time.

For now, if it doesn't suit a given use case, there are the finer-grained functions you can use.

File: src/k8s/core.clj (hunk 1 above)
Comment (ct7t) [reply src/k8s/core.clj:53]: Not sure if you're unsure about your comp order or mine - I verified mine's correct for my intent.  The main purpose for the surrounding function is because I only want the function applied if the key is a `:keyword`, otherwise I want the key left alone.  (There is an occasional string and this should be treated literally.)

Screen Shot 2021-05-05 at 3 35 20 PM

I tried various permutations of what you were suggesting.  None seemed to get me where I wanted to go...but also a little unsure why we'd try and replace `clojure.core/name` with an equivalent 3rd party function.
Screen Shot 2021-05-05 at 3 38 36 PM
Screen Shot 2021-05-05 at 3 37 32 PM
Screen Shot 2021-05-05 at 3 37 02 PM

Happy to clarify with a docstring.

Comment (joel7t) [reply src/k8s/core.clj:53]: If it's just keywords and not strings, then let's make it specific either in a docstring, or in the function name: `camel-case-keyword-keys`.

`inflections/camel-case-keys` will only work on the keys of maps, so it needs more than a keyword as input.

```clojure
user> (def m {:camel-says-moo 1 "dog-says-bark" 2})
#'user/m
user> ((comp #(inflections/camel-case-keys % :lower) inflections/stringify-keys) m)
{:camelSaysMoo 1, :dogSaysBark 2}
user> ((comp inflections/stringify-keys #(inflections/camel-case-keys % :lower)) m)
{"camelSaysMoo" 1, "dogSaysBark" 2}
```

In your case, you want the output to be `{"CamelSaysMoo" 1, "dog-says-bark" 2}`, if I understand correctly.  For maximum clarity, you could go with `camel-case-keyword-keys` and then comp with with `inflections/stringify-keys` (having removed the name call from your function), but that becomes long winded.  You could also go with `sanitize-keys` and elaborate in the docstring.

To summarize my concern: I see `camel-keys`, but the functionality only operates on keywords, and it also stringifies those keys, neither of which are obvious from the function name, which could lead to buggy code later.
